pkg/crypto/rings/client.go

```diff
@@ -194,3 +209,62 @@ func (rc *ringClient) addressesToPubKeys(
 	}
 	return pubKeys, nil
 }
+
+// getDelegationsAtBlock returns the appropriate delegations for the given
+// application and target block height.
+// The delegations selection is determined by the following:
+// - The delegations must be archived AFTER the provided block height.
+// - Among the delegations that meet the above criteria, the oldest one is chosen.
+// - If no suitable archived delegation is found, the active one is used.
+//
+// The illustration below shows the selection process:
+//
+// Legend:
+// x : Delegations archival
+// > : Active delegations
+// - : Activity time for a delegation set prior to its archival "x"
+// | : Target block height for which the delegations are requested
+// ^ : Selected delegations, archived "x" or active ">"
+//
+// Case: Picking an archived delegation
+// -----x---x-----x---x---->
+// -----------|---^
+//
+// Case: Picking the currently active delegation
+// -----x---x-----x---x---->
+// ---------------------|--^
+func getDelegationsAtBlock(
+	app *apptypes.Application,
+	blockHeight int64,
+) []string {
+	// neoProximateArchivedDelegations will hold the appropriate archived delegations
+	// for the provided block height or nil if no suitable one is found.
+	// It will be the oldest archived delegations among the ones that are past the
+	// provided block height.
+	var neoProximateArchivedDelegations []string
+	// neoProximateArchivalHeight will hold the block height of the appropriate
+	// archived delegation or math.MaxInt64 if no suitable one is found.
+	neoProximateArchivalHeight := int64(math.MaxInt64)
+
+	// The chronological order of the archived delegations may not be guaranteed
+	// and iterating through all of them is necessary to find the appropriate one.
+	for _, archivedDelegateeSet := range app.ArchivedDelegations {
+		// A valid archived delegation is one that is archived AFTER the provided height.
+		isArchivedAfterHeight := archivedDelegateeSet.LastActiveBlockHeight >= blockHeight
+		// And is the oldest archived set found so far.
+		isNeoProximate := archivedDelegateeSet.LastActiveBlockHeight < neoProximateArchivalHeight
+
+		if isArchivedAfterHeight && isNeoProximate {
+			neoProximateArchivalHeight = archivedDelegateeSet.LastActiveBlockHeight
+			neoProximateArchivedDelegations = archivedDelegateeSet.GatewayAddresses
+		}
+	}
+
+	// Return the archived delegations if one is found.
+	if neoProximateArchivedDelegations != nil {
+		return neoProximateArchivedDelegations
+	}
+
+	// Use the active one otherwise.
```

Comment:
I don't fully understand this. I'd argue we ALWAYS need to use `DelegateGatewayAddresses`. What am I missing?